Filter credentials sent by the native cred store on Build

simonferquel opened the pull request:
Previously, this used to fetch all credentials from the cred store
without any filtering. So if the underlying native credstore contained credentials
for things not related to Docker (which was the case with the mac and
windows credential helpers), this would send much more credentials to
the daemon than required.

Additionaly, if the user uses multiple config files for different
environments, credentials for every environment would be sent without
filtering to every targeted daemon.

Now, instead of asking the credential store a list of credentials, we
list all authenticated registries from the current config file, and we
ask the credential store if it has credentials for them.

fixes #31517 

**- What I did**

Filter which credentials are sent to the daemon on `docker build` from auths referenced in the config file

**- How I did it**

I removed the listCredentialsInStore method, and used the credentials list in the current config file to know which credentials should be sent

**- How to verify it**

Use Docker for mac or windows, patch docker cli with the one from this PR, 
- login to a private registry, 
- create a docker file with a FROM `something in this private registry`
- make sure docker build success
- edit the docker.config file and remove the corresponding auth entry
- make sure docker build fails

**- Description for the changelog**
Fixes an issue with credential helpers where too many credentials were sent to the daemon

simonferquel commented:
cc @jeanlaurent, @dgageot

thaJeztah commented:
Is this gonna work well with, e.g. the AWS or GCE authentication helpers? IIRC part of this was because those generate tokens on-the-fly, but I may be wrong

simonferquel commented:
@samuelkarp, could you have a look at this change and tell us if it would break things for you ?

vdemeester commented:
I also think it's gonna break the external `credStore` like the AWS and GCE ones (cc @KingEmet).
On that issue (too much credentials are sent), I really think the responsibilities was on `docker-credential-helpers` and that has been fixed (we now *label* them and listing them only).

/cc @docker/core-maintainers

simonferquel commented:
Hmm, understood the issue with AWS/GCR that implicitly provide credentials for the registries accessible by the user.

So could we change the meaning of `list` by `list implicitly authenticated registries` ? This way, a credential helper should only return the list of registries for which the user has never done a `docker login` call (for desktop cred helpers, it means that list returns nothing).

simonferquel committed:
Changes the expected behavior of credential helpers 'list' command

So that it returns only registries for which no `docker login` has been
called.

This enables to fix correctly the issue where windows and mac
credentials helper would send credentials unrelated to Docker to the
Daemon, while preservice GCR/AWS credential helpers functionality

This requires a change on the credential helpers for windows and mac so
that there list command returns an empty json object

Signed-off-by: Simon Ferquel <simon.ferquel@docker.com>

simonferquel commented:
@thaJeztah, @vdemeester, I amended the PR for changing the way list should behave : only return registries implicitly authenticated (ie:without `docker login`).

Does it make sense ?

KingEmet commented:
IIRC there were plans to eliminate creation of 'auths' entries once the email field was fully deprecated. If a cred store/helper exists, I don't believe there is currently any point for there to be an additional auths entry.

I think that the existing changes being added to docker-credential-helpers should solve the issues of superfluous keychain items being touched, but I believe that the core issue with `docker build` lies further upstream. As mentioned in #31517 I think the 'correct' solution would be to somehow limit the number of creds being requested to just those that are required for the build to succeed.

thaJeztah commented:
@KingEmet yes; the difficult bit is that image/registry resolution doesn't take place at the client, but daemon side (it's _possible_ to do at the client, but ideally, the client should not do that). I was brainstorming with @simonferquel for a solution where the daemon could request these credentials from the client (also see https://github.com/docker/docker/issues/31829#issuecomment-287068206)

samuelkarp commented:
@simonferquel, thanks for pinging me!

I don't think using the `auths` entries from `~/.docker/config.json` is the best answer for limiting credentials requested, but neither is calling `list` and asking the credential helper.  Either way, you're getting unintended credentials sent to the daemon.  Both of these are problematic for customers wanting to use the Amazon ECR Credential Helper.  Instead, parsing the `Dockerfile` and looking at the `FROM` line to get only the intended registry would (in my opinion) be the best mechanism.

For Amazon ECR, there's no way to enumerate all the registries that a particular entity (IAM user, IAM role) has access to.  Authorization is determined through a combination of IAM policies applied to the accessing entity as well as resource policies applied to the individual repositories.  IAM policies have a lot of flexibility to conditionally allow access based on access mechanisms (e.g., requiring MFA) as opposed to simply identity.  Additionally, because repositories can be accessed by entities in a separate account from the owning account, there's no efficient query mechanism that would list all repositories available to an accessing entity.  In the credential helper, we get around this today by caching all registries that have been requested for access (through `docker push` and `docker pull`) and just returning that list when `list` is called; this makes it hard to use `docker build` if you've never tried to access the registry that your `FROM` line references.  If `docker build` parsed the `Dockerfile` and looked at the `FROM` line, this problem would go away and you'd get the additional benefit of strongly scoping the set of credentials sent to the Docker daemon to be just the ones required for build.

simonferquel commented:
I'm closing the PR as it is incompatible with the way cloud credential helpers work
